### PR TITLE
Ensure client.restart waits for workers to leave and come back

### PR DESCRIPTION
**Pull request by fjetter**
Client should be blocking until the workers are cycled through. At the very least since https://github.com/dask/distributed/pull/6504 we're no longer removing workers immediately from the internal state tracking and therefore client.restart would return immediately before any workers were removed.

This is at the very least an issue since https://github.com/dask/distributed/pull/6504 possibly before. I have a strong suspicion that this is responsible for https://github.com/coiled/coiled-runtime/issues/166
I believe in cases of timeouts of individual nanny restarts, this was a problem before #6504 already

### CODE REVIEW

File: distributed/tests/test_client.py
Comment (hendrikmakait) [inline distributed/tests/test_client.py:3500]:
```suggestion
    await c.restart()
    assert len(s.workers) == len(initial_workers)
```

File: distributed/tests/test_client.py
Comment (fjetter) [reply distributed/tests/test_client.py:3501]: yes, the challenge here is to not test the implementation. We can, however, check that the workers we started with are no longer known to the scheduler

File: distributed/scheduler.py
Comment (fjetter) [reply distributed/scheduler.py:5148]: Well, the "correct" way would be to time out if this condition is not met and not silently return. That would be a breaking change and I'm not convinced this is worth it

File: distributed/tests/test_client.py
Comment (hendrikmakait) [reply distributed/tests/test_client.py]: I agree that we generally should not test the implementation, but I think in this case, it simply turns out that we can implement the functionality using the actual conditions we would like to test, so this should be fine.

File: distributed/scheduler.py
Comment (fjetter) [reply distributed/scheduler.py]: yes, and I think what we do is bad practice. We simply wait for 10 seconds and return regardless of whether the cluster is in the state we want it to be or not. Instead, we should raise a TimeoutError but that would be a breaking change.

Comment (hendrikmakait) [reply distributed/scheduler.py]: My bad, I read that as an `OR` condition which felt weird. Your explanation totally makes sense for and `AND`

**Review by hendrikmakait — APPROVED** [review]

**Review by hendrikmakait — COMMENTED** [review]
The new test appears to be broken in 2 CI runs, is that related or some CI GC issue?

**Comment by fjetter** [comment]
waiting for merge until that one is merged and I can rebase this PR

**Comment by fjetter** [comment]
The new test is unfortunately not solid. It appears that the workers are timing out

File: distributed/scheduler.py
Comment (gjoseph92) [inline distributed/scheduler.py:5169]: See this comment: https://github.com/dask/distributed/pull/6637/files#diff-bbcf2e505bf2f9dd0dc25de4582115ee4ed4a6e80997affc7b22122912cc6591R5108-R5109

Workers without nannies will be closed and not come back. So we shouldn't wait for them. We should just be waiting for `nanny_workers`.

Comment (gjoseph92) [inline distributed/scheduler.py:5169]: Isn't a restarted worker going to have a new `server_id`? The ID is coming from the `Worker`, not the `Nanny` instance: https://github.com/dask/distributed/blob/d88c1d236da65f036af6ea05881c84f6b4c9fbb8/distributed/core.py#L341
https://github.com/dask/distributed/blob/d88c1d236da65f036af6ea05881c84f6b4c9fbb8/distributed/worker.py#L1113

So I don't know why we'd even expect those IDs to come back. Indeed, if they _did_ come back that would be highly concerning.

I think this might be reasonable (though obviously not tolerant to new workers joining during the restart):
```suggestion
            len(self.running) < len(nanny_workers)
```

Comment (hendrikmakait) [reply distributed/scheduler.py:5169]: The way I understood the intent of that comment and the code, we purposefully rely on the deployment system to bring the workers back up and we therefore want to wait for that to happen. Otherwise, in a non-nanny environment, `Scheduler.restart` would just be a Scheduler.shutdown` with extra steps. I might be wrong though...and we should definitely improve the way we handle the case where this does not happen.

Comment (hendrikmakait) [reply distributed/scheduler.py:5169]: Is there a scenario where due to some race condition an old worker might somehow still be in `self.worker` or reconnecting before shutting down? I think that's what this is intended to catch.

Comment (gjoseph92) [reply distributed/scheduler.py:5169]: Yes, that's possible if we hit the `not all(resp == "OK" for resp in resps)` branch when telling Nanny workers to restart. If the nanny times out trying to kill the worker subprocess, that worker may still be connected.

https://github.com/dask/distributed/blob/f7f650154fea29978906c65dd0225415da56ed11/distributed/nanny.py#L479-L491

However, in that case, I think we'd want to ignore that worker in our count of workers we expect to come back. Since we didn't successfully shut it down, it's still around, and if/when it does shut down in the future, it's quite possible it won't come back. We can disregard the fact that we tried to restart that one; it's a lost cause.

Honestly, I wonder if we should call `remove_worker` on the workers that didn't respond `"OK"`. Their state is a bit undefined, but it's probably kinda broken. Maybe we don't want to talk to them anymore. If we remove them, the scheduler will reject them the next time they try to talk to it, so it's safe.

But if we don't do that, then I think `initial_workers & self._get_worker_ids()` is too strict. Because the only reason some `initial_workers` would still be connected is that their nannies failed to shut down. If they failed to shut down, then they may never restart, so we shouldn't wait for them.

I'd then be torn on `len(self.workers) < n_workers` vs `len(self.running) < n_workers`. It's quite possible these broken workers aren't running anymore, and never will be. So just checking `len(self.workers)` means they'll still be counted. OTOH, it's nice to wait for new workers to be fully running instead of just connected. We could do
```python
                if n_restart_failed := sum(resp != "OK" for resp in resps):
                    logger.error(
                        "Not all workers responded positively: %s",
                        resps,
                        exc_info=True,
                    )
                    n_workers -= n_restart_failed
```
Then checking `len(self.running) < n_workers` will ignore the workers we don't expect to restart. But I don't know if it's necessary.

Comment (hendrikmakait) [reply distributed/scheduler.py:5169]: This is related to Florian's comment here: https://github.com/dask/distributed/pull/6637#discussion_r909503067

**Pull request closed without merging**
